### PR TITLE
fix(Datastore): Consecutive Updates (Save, Sync, Update and Immediately Delete Scenario)

### PR DESCRIPTION
**Pull request by thisisabhash**
*Issue #, if available:* #1206

*Description of changes:* This PR contains fix for Consecutive Update (Save, Sync, Update and Immediately Delete Scenario). The head of mutation event table for a given model id is reconciled with the version of the model in the API response after the models in the API request and response are compared for equality.

*Check points:*

- [x] Added new tests to cover change, if needed
- [x] All unit tests pass
- [x] All integration tests pass
~~- [ ] Documentation update for the change if required~~
- [x] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+// swiftlint:disable syntactic_sugar
+extension ModelSchema {
+
+    /// Compare two `Model` based on a given `ModelSchema`
+    /// Returns true if equal, false otherwise
+    public static func isEqual(_ model1: Model,
+                               _ model2: Model,
+                               _ modelSchema: ModelSchema) -> Bool {
+        for (fieldName, modelField) in modelSchema.fields {
+            // skip read only fields
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:21]: good place to add comment to why we skip read-only fields; like `updatedAt` is created/updated by the service so for the purpose of this equality check, we ignore it. what are the read-only fields other than updatedAt? does it hold true for those other fields as well?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+            let value1 = model1[fieldName] ?? nil
+            let value2 = model2[fieldName] ?? nil
+
+            // check equality for different `ModelFieldType`
+            switch modelField.type {
+            case .string:
+                guard let value1Optional = value1 as? String?, let value2Optional = value2 as? String? else {
+                    return false
+                }
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:34]: what if we have a `modelField.type` and then `value1` and `value2` are both `nil`- meaning both values do not exist on either models. it should continue because there is nothing to check but this returns false and says the model's aren't equal

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift [hunk 3]
```diff
@@ -30,21 +29,34 @@ extension MutationEvent {
...
-                        """)
-                    existingEvent.version = mutationSync.syncMetadata.version
-                    storageAdapter.save(existingEvent, condition: nil) { result in
-                        switch result {
-                        case .failure(let dataStoreError):
-                            completion(.failure(dataStoreError))
-                        case .success:
-                            completion(.success(()))
+                // check if version returned from API is not nil and <= version sent in the request
+                if existingEvent.version != nil && mutationSync.syncMetadata.version <= existingEvent.version! {
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:33]: 1. if the `mutationSync.syncMetadata.version` is == to `existingEvent.version` then if the model equality check checks out, we replace it anyways- so i think it should just execute the below logic for `<`.
2. if `existing.version == nil` it continues to the logic below. before it would replace it regardless of what the response version is, but now we check the model. do we want to do this? if the existing.version is `nil`, can we avoid the model check and simply replace it?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift [hunk 4]
```diff
@@ -30,21 +29,34 @@ extension MutationEvent {
...
+                    return
+                }
+
+                do {
+                    let responseModel = mutationSync.model.instance
+                    let requestModel = try mutationEvent.decodeModel()
+
+                    // check if the data sent in the request is the same as the response
+                    // if it is, update the pending mutation event version to the response version
+                    if let modelSchema = ModelRegistry.modelSchema(from: mutationEvent.modelName),
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:44]: nit: inverse with `guard` instead of if else

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginIntegrationTests/DataStoreConsecutiveUpdatesTests.swift [hunk 5]
```diff
@@ -292,9 +285,8 @@ class DataStoreConsecutiveUpdatesTests: SyncEngineIntegrationTestBase {
 
             if mutationEvent.mutationType == GraphQLMutationType.delete.rawValue {
                 XCTAssertEqual(post, updatedPost)
-                // this needs to be commented out once the bug is fixed
-                // deleteSyncReceived.fulfill()
-                // XCTAssertEqual(mutationEvent.version, 3)
+                deleteSyncReceived.fulfill()
+                XCTAssertEqual(mutationEvent.version, 3)
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginIntegrationTests/DataStoreConsecutiveUpdatesTests.swift:289]: nice!

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift [hunk 6]
```diff
@@ -30,21 +29,34 @@ extension MutationEvent {
...
-                            \(mutationSync.syncMetadata.version)
-                        """)
-                    existingEvent.version = mutationSync.syncMetadata.version
-                    storageAdapter.save(existingEvent, condition: nil) { result in
-                        switch result {
-                        case .failure(let dataStoreError):
-                            completion(.failure(dataStoreError))
-                        case .success:
-                            completion(.success(()))
+                // check if version returned from API is not nil and <= version sent in the request
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:32]: `// check if the pending mutation version is not nil and <= version from the response``

Comment (lawmicha) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:32]:
```
guard let existingVersion = existingEvent.version, existingVersion <= mutationSync.syncMetadata.version else {
completion(.success(()))
   return
}
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift (hunk 3 above)
Comment (thisisabhash) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:33]: `if existingEvent.version!=nil && existingEvent.version >= mutationSync.syncMetadata.version`

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift [hunk 7]
```diff
@@ -30,21 +29,34 @@ extension MutationEvent {
...
+                }
+
+                do {
+                    let responseModel = mutationSync.model.instance
+                    let requestModel = try mutationEvent.decodeModel()
+
+                    // check if the data sent in the request is the same as the response
+                    // if it is, update the pending mutation event version to the response version
+                    if let modelSchema = ModelRegistry.modelSchema(from: mutationEvent.modelName),
+                       ModelSchema.isEqual(responseModel, requestModel, modelSchema) {
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:45]: `modelSchema.isEqual` since there is an extension on modelSchema instances

Comment (lawmicha) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:45]:
```
guard let modelSchema = ModelRegistry.modelSchema(from: mutationEvent.modelName), modelSchema.isEqual(model1, model1) else {
  complete(.success())
}
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+// swiftlint:disable cyclomatic_complexity
+// swiftlint:disable syntactic_sugar
+extension ModelSchema {
+
+    /// Compare two `Model` based on a given `ModelSchema`
+    /// Returns true if equal, false otherwise
+    public static func isEqual(_ model1: Model,
+                               _ model2: Model,
+                               _ modelSchema: ModelSchema) -> Bool {
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:19]: `public func isEqual`

Comment (lawmicha) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift]: my mistake, this probably doesn't need to be `public`

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift [hunk 9]
```diff
@@ -199,4 +239,317 @@ class MutationEventExtensionsTest: BaseDataStoreTests {
...
+        wait(for: [updatingVersionExpectation], timeout: 1)
+
+        // query for head of mutation event table for given model id and check if it has the correct version
+        MutationEvent.pendingMutationEvents(for: post1.id,
+                                            storageAdapter: storageAdapter) { result in
+            switch result {
+            case .failure(let error):
+                XCTFail("Error : \(error)")
+            case .success(let mutationEvents):
+                guard !mutationEvents.isEmpty, let updatedEvent = mutationEvents.first else {
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift:439]: if it wasn't supposed to be updated, `updatedEvent` is confusing, perhaps just `pendingMutationEvent` or `mutationEventHead`

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift [hunk 10]
```diff
@@ -11,15 +11,14 @@ import AWSPluginsCore
...
-    // if it has a `nil` version, with syncMetadata version in `mutationSync`
-    // and saves it in the mutation event table
-    static func updatePendingMutationEventVersionIfNil(for modelId: Model.Identifier,
-                                                       mutationSync: MutationSync<AnyModel>,
-                                                       storageAdapter: StorageEngineAdapter,
-                                                       completion: @escaping DataStoreCallback<Void>) {
+    // Updates the version of the head of pending mutation event queue for a given model `id`
+    // with syncMetadata version in `mutationSync` and saves it in the mutation event table
+    static func reconcilePendingMutationEventsVersion(mutationEvent: MutationEvent,
+                                                      mutationSync: MutationSync<AnyModel>,
```
[the remark this answers is not in the document]
Comment (thisisabhash) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:17]: the test `testRecordCoversAreEqualWithDifferentReadOnlyFields` tests the read only fields case where `createdAt` is a read only field.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/ModelCompareTests.swift [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+import SQLite
+import XCTest
+
+@testable import Amplify
+@testable import AmplifyTestCommon
+@testable import AWSDataStoreCategoryPlugin
+@testable import AWSPluginsCore
+
+// swiftlint:disable type_body_length
+class ModelCompareTests: BaseDataStoreTests {
```
[the remark this answers is not in the document]
Comment (thisisabhash) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/ModelCompareTests.swift:18]: the test `testRecordCoversAreEqualWithDifferentReadOnlyFields` tests the read only fields case where `createdAt` is a read only field.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/ModelCompareTests.swift [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+        let color1 = Color(name: "Color1", red: 100, green: 100, blue: 100)
+        let color2 = Color(name: "Color2", red: 200, green: 200, blue: 200)
+        let category1 = Category(name: "Category1", color: color1)
+        let category2 = Category(name: "Category1", color: color2)
+        let todo1 = Todo(id: id, name: name, categories: [category1])
+        let todo2 = Todo(id: id, name: name, categories: [category2])
+        XCTAssertFalse(Todo.schema.isEqual(todo1, todo2))
+    }
+
+    func testRecordCoversAreEqualWithDifferentReadOnlyFields() {
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/ModelCompareTests.swift:305]: perhaps a test comment that "createdAt" is a read-only field

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+                            continue
+                        }
+                    }
+                } catch {
+                    continue
+                }
+            case .model:
+                continue
+            case .collection:
+                continue
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:137]: explain why these are skipped

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/ModelCompareTests.swift [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+import SQLite
+import XCTest
+
+@testable import Amplify
+@testable import AmplifyTestCommon
+@testable import AWSDataStoreCategoryPlugin
+@testable import AWSPluginsCore
+
+// swiftlint:disable type_body_length
+class ModelCompareTests: BaseDataStoreTests {
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/ModelCompareTests.swift:18]: can you add test for when a Post has a `nil` field value and the other Post has a value? they are not equal, right?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+            // check equality for different `ModelFieldType`
+            switch modelField.type {
+            case .string:
+                guard let value1Optional = value1 as? String?, let value2Optional = value2 as? String? else {
+                    return false
+                }
+                if let value1 = value1Optional, let value2 = value2Optional, value1 != value2 {
+                    return false
+                } else {
+                    continue
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:36]: if `value1Optional` cannot be unwrapped because it's `nil` but value2Optional could have been unwrapped, then we should return false, right? since one side is nil and the other side is valid, but this logic would continue and move onto the next field

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift [hunk 16]
```diff
@@ -199,4 +239,317 @@ class MutationEventExtensionsTest: BaseDataStoreTests {
...
+                                                        modelId: modelId,
+                                                        modelName: Post.modelName,
+                                                        json: try post2.toJSON(),
+                                                        mutationType: .update,
+                                                        version: 2)
+
+        let metadata = MutationSyncMetadata(id: modelId,
+                                            deleted: false,
+                                            lastChangedAt: Int(Date().timeIntervalSince1970),
+                                            version: 1)
```
[the remark this answers is not in the document]
Comment (thisisabhash) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift:269]: This test is for when we receive an older version in the response and the version in pending mutation event table is newer.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift [hunk 17]
```diff
@@ -187,16 +227,320 @@ class MutationEventExtensionsTest: BaseDataStoreTests {
...
+        }
+
+        let pendingMutationEventExpectation = expectation(description: "save pendingMutationEvent success")
+        storageAdapter.save(pendingMutationEvent) { result in
+            guard case .success = result else {
+                XCTFail("Failed to save metadata")
+                return
+            }
+            pendingMutationEventExpectation.fulfill()
+        }
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift:289]: this can be reduced to a set up step that invovles saving an array of pending mutation events, and verifying that the head is as expected:
```
setUpPendingMutationQueue([requestMutationEvent, pendingMutationEvent])
```


```
func setUpPendingMutationQueue(_ mutationEvents: [MutationEvent]) {
   // for each mutationEvent, perform storageAdapter.save

  // verify the head of the queue is as expected
}
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift [hunk 18]
```diff
@@ -30,11 +35,25 @@ extension MutationEvent {
...
+
+                    // check if the data sent in the request is the same as the response
+                    // if it is, update the pending mutation event version to the response version
+                    guard let modelSchema = ModelRegistry.modelSchema(from: mutationEvent.modelName),
+                          modelSchema.isEqual(responseModel, requestModel) else {
+                        completion(.success(()))
+                        return
+                    }
+
                     existingEvent.version = mutationSync.syncMetadata.version
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:57]: can this be factored out so it can be called like this? (and easier to test)

```swift
guard let reconciledEvent = reconcile(pendingMutationEvent: existingEvent, with: mutationEvent, responseMutationSync: mutationSync) else {
  complete(.success())
 return
}

 storageAdapter.save(reconciledEvent)
```
```swift
static func reconcile(pendingMutationEvent: MutationEvent, 
         with requestMutationEvent: MutationEvent, 
          responseMutationSync: MutationSync<AnyModel>) -> MutationEvent? {


 // return if version of the pending mutation event is not nil and
                // is >= version contained in the response
                if pendingMutationEvent.version != nil && pendingMutationEvent.version! >= mutationSync.syncMetadata.version {
                    return nil
                }

                do {
                    let responseModel = mutationSync.model.instance
                    let requestModel = try mutationEvent.decodeModel()

                    // check if the data sent in the request is the same as the response
                    // if it is, update the pending mutation event version to the response version
                    guard let modelSchema = ModelRegistry.modelSchema(from: mutationEvent.modelName),
                          modelSchema.isEqual(responseModel, requestModel) else {
                        return nil
                    }

                    pendingMutationEvent.version = mutationSync.syncMetadata.version
                 return pendingMutationEvent
          } catch {
               Amplify.log.verbose("Error decoding models: \(error)")
               return nil
          }
}


existing unit tests can be extended with one line:
```swift
// currently we unit test `reconcilePendingMutationEventsVersion` like so, followed by verifying the pending mutation queue
MutationEvent.reconcilePendingMutationEventsVersion(sent: request,
                                                             received: response,
                                                             storageAdapter: storageAdapter) 

// you can add an additional check like so:
let reconciledEvent = reconcile(pendingMutationEvent: head, with: mutationEvent, responseMutationSync: response)

// depending on the situation you would either
XCTAssertNil(reconciledEvent) // no version update
// or
XCTAssertEqual(reconciledEvent.version, response.version)
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift [hunk 19]
```diff
@@ -14,26 +14,49 @@ import XCTest
 @testable import AWSDataStoreCategoryPlugin
 @testable import AWSPluginsCore
 
+// swiftlint:disable cyclomatic_complexity
+// swiftlint:disable type_body_length
+// swiftlint:disable file_length
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift:19]: added a few suggestions below to reduce file length

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+                if !modelCompareHelper(value1Optional, value2Optional) {
+                    return false
+                }
+            case .int:
+                guard let value1Optional = value1 as? Int?, let value2Optional = value2 as? Int? else {
+                    return false
+                }
+                if !modelCompareHelper(value1Optional, value2Optional) {
+                    return false
+                }
```
[the remark this answers is not in the document]
Comment (thisisabhash) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:42]: it can't be a return statement as in true case, we need to continue checking through other fields

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift [hunk 21]
```diff
@@ -10,16 +10,21 @@ import Dispatch
...
-
-    // Updates the head of pending mutation event queue for a given model `id`
-    // if it has a `nil` version, with syncMetadata version in `mutationSync`
-    // and saves it in the mutation event table
-    static func updatePendingMutationEventVersionIfNil(for modelId: Model.Identifier,
-                                                       mutationSync: MutationSync<AnyModel>,
-                                                       storageAdapter: StorageEngineAdapter,
-                                                       completion: @escaping DataStoreCallback<Void>) {
+    // Reconciles the version of pending mutation events in mutation event table with the
+    // API response version for the following consecutive update scenarios :
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:14]: why do we reconcile? Consecutive operations that modify a model results in a sequence of pending mutation events that have the current version of the model. The first mutation event has the correct version of the model, while the subsequent events will have lower versions if the first mutation event is synced to the cloud. By checking reconciling the pending mutation events after syncing the first mutation event, we attempt to update the pending version to the latest version of the object.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift [hunk 22]
```diff
@@ -10,16 +10,21 @@ import Dispatch
...
+    //  - Save, then update
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift:15]: "Save, then immediately update" makes more sense, for each scenario there are versions involves, what does the mutation queue look like and what are the versions for each operation? what does reconcile do in this scenario?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift [hunk 23]
```diff
@@ -88,69 +123,115 @@ class MutationEventExtensionsTest: BaseDataStoreTests {
...
-                                                        version: nil)
-        let metadata = MutationSyncMetadata(id: modelId,
-                                            deleted: false,
-                                            lastChangedAt: Int(Date().timeIntervalSince1970),
-                                            version: 1)
+        let post1 = Post(id: modelId, title: "title1", content: "content1", createdAt: .now())
+        let post2 = Post(id: modelId, title: "title2", content: "content2", createdAt: .now())
+        let requestMutationEvent =
+            try createMutationEvent(model: post1, mutationType: .create, version: 2, inProcess: true)
+        let pendingMutationEvent = try createMutationEvent(model: post2, mutationType: .update, version: 2)
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift:150]: nit: remove mutationType parameter if it doesn't matter

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift [hunk 24]
```diff
@@ -88,69 +123,115 @@ class MutationEventExtensionsTest: BaseDataStoreTests {
...
+        let responseMutationSync = createMutationSync(model: post1, deleted: false, version: 1)
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift:151]: nit: remove deleted parameter if it doesn't matter

**Review by lawmicha — APPROVED** [review]

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift (hunk 23 above)
Comment (thisisabhash) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginTests/Sync/Support/MutationEventExtensionsTests.swift:150]: type is needed for `testSentModelWithNilVersion_SecondPendingEventNotReconciled` where we have `update` and `delete` events in queue and only `update` event is modified. Will let it be as the tests model the different scenarios like "Save, then update" etc.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+
+// swiftlint:disable cyclomatic_complexity
+extension ModelSchema {
+
+    private static let serviceUpdatedFields: Set = ["updatedAt"]
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:14]: - can use a `Set` for faster look up
- is it possible to add `createdAt` as an explicit and optional field? resulting in `createdAt` not a read-only and optionally passed in when creating the model. if this case is true, then should add `createdAt` into the Set as well

Comment (thisisabhash) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:14]: it seems logical to add `createdAt` to the list - will do.
is there a faster `Set` implementation?

**Review by lawmicha — APPROVED** [review]

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+
+// swiftlint:disable cyclomatic_complexity
+extension ModelSchema {
+
+    private static let serviceUpdatedFields: Set = ["updatedAt", "createdAt"]
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:14]: add comment regarding this

Comment (lawmicha) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:14]: which cases we fix, which we don't

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+
+    /// Compare two `Model` based on a given `ModelSchema`
+    /// Returns true if equal, false otherwise
+    /// Currently, schemas where system timestamps fields (createdAt, updatedAt)
+    /// are changed with `timestamps` attribute to model directive and also added
+    /// to model schema are not covered. This will cause `isEqual` to fail in
+    /// such cases.
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:21]:
```suggestion
    /// Currently, schemas where system timestamps fields (createdAt, updatedAt)
    /// are renamed using with `@model`'s `timestamps` attribute and explicitly
    /// added to the input schema are not supported by this check since they are
    /// marked as "read-only" fields and will fail the check when the service generates
    /// and returns the value of `createdAt` or `updatedAt`.
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+extension ModelSchema {
+
+    private static let serviceUpdatedFields: Set = ["updatedAt", "createdAt"]
+
+    /// Compare two `Model` based on a given `ModelSchema`
+    /// Returns true if equal, false otherwise
+    /// Currently, schemas where system timestamps fields (createdAt, updatedAt)
+    /// are renamed using with `@model`'s `timestamps` attribute and explicitly
+    /// added to the input schema are not supported by this check since they are
+    /// marked as "read-only" fields and will fail the check when the service generates
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:21]:
```suggestion
    /// added to the input schema are not supported by this check since they are not
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+// swiftlint:disable cyclomatic_complexity
+extension ModelSchema {
+
+    private static let serviceUpdatedFields: Set = ["updatedAt", "createdAt"]
+
+    /// Compare two `Model` based on a given `ModelSchema`
+    /// Returns true if equal, false otherwise
+    /// Currently, schemas where system timestamps fields (createdAt, updatedAt)
+    /// are renamed using with `@model`'s `timestamps` attribute and explicitly
+    /// added to the input schema are not supported by this check since they are
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift:20]:
```suggestion
    /// added to the input schema are not supported by this check since they are not
```

**Review by lawmicha — APPROVED** [review]
